glusterd2/commands/snapshot/snapshot-restore.go

```diff
@@ -0,0 +1,375 @@
+		newVol.Options[key] = value
+	}
+
+	newVol.SnapList = vol.SnapList
+	newVol.State = vol.State
+	newVol.Transport = snapVol.Transport
+	newVol.Type = snapVol.Type
+	newVol.VolfileID = newVol.Name
+	for idx, subvol := range snapVol.Subvols {
+		subvolType := volume.SubvolTypeToString(subvol.Type)
+		name := fmt.Sprintf("%s-%s-%d", vol.Name, strings.ToLower(subvolType), idx)
+		bricks := createVolumeBrickFromSnap(subvol.Bricks, vol)
+		s := volume.Subvol{
+			ArbiterCount:    subvol.ArbiterCount,
+			DisperseCount:   subvol.ArbiterCount,
+			ID:              subvol.ID,
+			Name:            name,
+			RedundancyCount: subvol.RedundancyCount,
+			ReplicaCount:    subvol.ReplicaCount,
+			Type:            subvol.Type,
+			Subvols:         subvol.Subvols,
+			Bricks:          bricks,
+		}
+		newVol.Subvols = append(newVol.Subvols, s)
+		/*
+			TODO
+			*Checksum
+			*newVol.Version = snapVol.Version
+		*/
+	}
+	return newVol
+}
+
+func storeSnapRestore(c transaction.TxnCtx) error {
+	var snapname string
+	if err := c.Get("snapname", &snapname); err != nil {
+		return err
+	}
+
+	snapInfo, err := snapshot.GetSnapshot(snapname)
+	if err != nil {
+		return err
+	}
+	snapVol := &snapInfo.SnapVolinfo
+
+	vol, err := volume.GetVolume(snapInfo.ParentVolume)
+	if err != nil {
+		return err
+	}
+
+	newVolinfo := createRestoreVolinfo(snapInfo, vol)
+
+	if err := volume.AddOrUpdateVolumeFunc(&newVolinfo); err != nil {
+		c.Logger().WithError(err).WithField(
+			"volume", newVolinfo.Name).Debug("failed to store volume info")
+		return err
+	}
+
+	if err := volgen.Generate(); err != nil {
+		c.Logger().WithError(err).WithField(
+			"volume", newVolinfo.Name).Debug("failed to generate volfiles")
+		return err
+	}
+	if err := snapshot.DeleteSnapshot(snapInfo); err != nil {
+		c.Logger().WithError(err).WithField(
+			"snapshot", snapVol.Name).Debug("failed to delete snap from store")
+	}
+
+	return nil
+}
+
+func registerSnapRestoreStepFuncs() {
+	var sfs = []struct {
+		name string
+		sf   transaction.StepFunc
+	}{
+		{"snap-restore.Commit", snapRestore},
+		{"snap-restore.UndoCommit", undoSnapRestore},
+		{"snap-restore.UndoStore", undoSnapStore},
+		{"snap-restore.Store", storeSnapRestore},
+	}
+	for _, sf := range sfs {
+		transaction.RegisterStepFunc(sf.sf, sf.name)
+	}
+}
+
+func snapshotRestoreHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	logger := gdctx.GetReqLogger(ctx)
+	snapname := mux.Vars(r)["snapname"]
+
+	snapinfo, err := snapshot.GetSnapshot(snapname)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+
+	txn, err := transaction.NewTxnWithLocks(ctx, snapname, snapinfo.ParentVolume)
+	if err != nil {
+		status, err := restutils.ErrToStatusCode(err)
+		restutils.SendHTTPError(ctx, w, status, err)
+		return
+	}
+	defer txn.Done()
+
+	//Fetching snapinfo again, but this time inside a lock
+	snapinfo, err = snapshot.GetSnapshot(snapname)
```

Comment:
Snapinfo is a modifiable object right? Then wouldn't you need a lock on the snapname before fetching and using it? Maybe take a lock on "snap-"+snapname.

Reply:
done